docs/proposals/vip-3.md

```diff
@@ -0,0 +1,226 @@
+this project and operating the end result safely.
+
+In order to write idiomatic, higher quality, Rust code, the project will adopt [rustfmt](https://github.com/rust-lang/rustfmt) 
+and [rust-clippy](https://github.com/rust-lang/rust-clippy).
+
+### Why gRPC?
+
+Because it seems easier than to re-implement Venice's bespoke [REST spec](../dev_guide/router_rest_spec.md). Also,
+although gRPC is probably not perfect, it is likely more efficient than the current REST protocol, and will make it
+easier to support clients in other languages. Also, we already have gRPC embedded in the stack following the Summer 2023
+internship project, so it might as well be leveraged.
+
+There is more than one crate option for integrating with gRPC, but the most mature one seems to be [Tonic](https://crates.io/crates/tonic).
+
+### Why JNI?
+
+Because it is supported in Rust, and that should make the integration easier. An alternative would be to have the Java 
+server communicate with the Rust server via some other mechanism (loopback socket, pipes, etc) but then that would mean
+adding lots of forks in the code, and might also be less efficient. Using a non-JNI mechanism would also mean that just
+enabling the Rust server (even if it's not used yet) would already alter the behavior of the main code, thus creating
+more risk. In order to work in an incremental manner, where the Rust code can be developed in small chunks and hidden
+behind a config flag until ready, it is preferable to keep this interface the same.
+
+In the end state, JNI would only be used by the Java code responsible for the ingestion, and the read path would be
+fully JNI-free. Also, it should be noted that the Rust server will not be a separate child process spawned by the Java
+process, but rather fully embedded in the Java process. The Rust server will be activated by a JNI call initiated by the
+Java server, which spins up the gRPC Handler on the Rust side, along with any port and threads it needs to function.
+
+Doing JNI properly requires packaging a jar with code compiled for multiple architectures (at least Mac and Linux for
+the time being). There is some complexity there, but it can be done.
+
+### Quota
+
+The read quota will need to be re-implemented in Rust. There are essentially two main approaches we can consider:
+
+1. Let the Rust server and Java server share a quota amount between them, so that no matter if the traffic comes in via
+   the Java port or the Rust port, it will be accounted for and the configured limit will be respected. This requires
+   having some kind of synchronization between the two sides (either at the granularity of each request, which is likely
+   too expensive, or by reserving some allotment periodically, which is complicated).
+2. Let the Rust server and Java server each have the full quota amount and not need any interaction between each other 
```

Comment:
I assume the initial version of Rust server will only support single-get/batch-get, which doesn't require complex handling logic.
Maybe in the first phase, we only allow the users, which are not using read compute to access Rust Server, so that Rust Server will take all the traffic.
Once Rust Server supports read compute, we can migrate these leftover use cases.

Reply:
Yes, it would be convenient to run only one implementation at a time... I think it is a good idea.